### PR TITLE
Add staking and rewards for wallet addresses

### PR DESCRIPTION
**Pull request by marcin-cb**
### What changed? Why?
Adding the ability for wallet addresses to perform staking operations and view rewards.
This feature will only be supported for Partial ETH staking at this time.

The interfaces that are implemented:
#### Wallet functions
```
# create a staking operation
let stakingOperation  = await wallet.createStake(0.0001, "eth");

# get the stakeable balance
let stakeableBalance = await wallet.stakeableBalance("eth")

# get the unstable balance
let unstakeableBalance = await wallet.unstakeableBalance("eth");

# get the claimable balance
let claimableBalance = await wallet.claimableBalance("eth")

# get the rewards (default is last 7 days)
let rewards = await wallet.stakingRewards("eth")
```

#### Address functions
```
# create a staking operation 
let stakingOperation = await wallet.getDefaultAddress()!.createStake(0.0001, "eth");

# get the stakeable balance
let stakeableBalance = await wallet.getDefaultAddress()!.stakeableBalance("eth")

# get the unstable balance
let unstakeableBalance = await wallet.getDefaultAddress()!.unstakeableBalance("eth");

# get the claimable balance
let claimableBalance = await wallet.getDefaultAddress()!.claimableBalance("eth")

# get the rewards (default is last 7 days)
let rewards = await wallet.getDefaultAddress()!.stakingRewards("eth")
```

### CODE REVIEW

**Comment by drohit-cb** [comment]
Can you update PR description with example outputs of the api being exposed.

File: src/coinbase/address/wallet_address.ts [hunk 1]
```diff
@@ -322,22 +337,266 @@ export class WalletAddress extends Address {
...
+   */
+  public async createStakingOperation(
+    amount: Amount,
+    assetId: string,
+    action: string,
+    timeoutSeconds = 60,
+    intervalSeconds = 0.2,
+    options: CoinbaseWalletAddressStakeOptions = { mode: StakeOptionsMode.DEFAULT },
+  ): Promise<StakingOperation> {
+    await this.validateAmount(amount, assetId);
```
Comment (drohit-cb) [inline src/coinbase/address/wallet_address.ts:378]: hmm should we rely on getBalance or should we depend on stakeableBalance ?

File: src/coinbase/address/wallet_address.ts [hunk 2]
```diff
@@ -322,22 +337,266 @@ export class WalletAddress extends Address {
...
+
+  /**
+   * Get the stakeable balance for the supplied asset.
+   *
+   * @param asset_id - The asset to check the stakeable balance for.
+   * @param mode - The staking mode. Defaults to DEFAULT.
+   * @param options - Additional options for getting the stakeable balance.
+   * @returns The stakeable balance.
+   */
+  public async stakeableBalance(
```
Comment (drohit-cb) [inline src/coinbase/address/wallet_address.ts:418]: can we move these calls common b/w the 2 models to the parent class address.ts ?

File: src/coinbase/address/wallet_address.ts [hunk 3]
```diff
@@ -322,22 +337,270 @@ export class WalletAddress extends Address {
...
+   *
+   * @param amount - The amount for the staking operation.
+   * @param assetId - The asset to the staking operation.
+   * @param action - The type of staking action to perform.
+   * @param timeoutSeconds - The amount to wait for the transaction to complete when broadcasted.
+   * @param intervalSeconds - The amount to check each time for a successful broadcast.
+   * @param options - Additional options such as setting the mode for the staking action.
+   * @returns The staking operation after it's completed fully.
+   */
+  public async createStakingOperation(
```
Comment (drohit-cb) [inline src/coinbase/address/wallet_address.ts:370]: any reason we don't expose a action specific method such as `createStakeOperation` similar to external address ?

File: src/coinbase/address/wallet_address.ts (hunk 1 above)
Comment (marcin-cb) [reply src/coinbase/address/wallet_address.ts:378]: Yeah good call out. That would be more accurate.

File: src/coinbase/address/wallet_address.ts (hunk 2 above)
Comment (deangalvin-cb) [reply src/coinbase/address/wallet_address.ts:420]: +1

File: src/coinbase/staking_operation.ts [hunk 4]
```diff
@@ -36,14 +37,76 @@ export class StakingOperation {
...
+    return this.model.address_id;
+  }
+
+  /**
+   * Returns the Status of the StakingOperation.
+   *
+   * @returns The Status of the StakingOperation.
+   */
+  public getStatus(): StakingOperationStatus | undefined {
+    switch (this.model.status) {
```
Comment (deangalvin-cb) [inline src/coinbase/staking_operation.ts:72]: Is this the best way to do this in typescript?

Comment (deangalvin-cb) [reply src/coinbase/staking_operation.ts:72]: Why not use status directly?

**Comment by amethystens** [comment]
Jak to zrobić? Masz możliwość zrobić to za mnie?

a.d.82 ᛉ

pt., 26 lip 2024, 15:02 użytkownik marcin-cb ***@***.***>
napisał:

**Review by amethystens — APPROVED** [review]

**Review by amethystens — APPROVED** [review]

**Review by amethystens — APPROVED** [review]

**Comment by amethystens** [comment]
a.d.82 ᛉ

pt., 26 lip 2024, 15:34 użytkownik Czak Bery ***@***.***> napisał:

**Comment by deangalvin-cb** [comment]
Hey @amethystens! We will have developer documentation on how to get started here: (soon) https://docs.cdp.coinbase.com/staking/docs/quickstart

We are also always available on our discord! https://discord.com/invite/cdp/

**Comment by amethystens** [comment]
Wszystko pięknie. Na tym smartfoniku nic nie zrobię. Myślę o nowym
sprzęcie, ale na tą chwilę tylko mogę myśleć. Chciałbym brnąc dalej bo to
wszystko wygląda bajecznie. Chyba muszę zrezygnować z suba. 😢

a.d.82 ᛉ

pt., 26 lip 2024, 16:15 użytkownik deangalvin-cb ***@***.***>
napisał:

File: src/coinbase/address/wallet_address.ts (hunk 2 above)
Comment (marcin-cb) [reply src/coinbase/address/wallet_address.ts]: Good point, done!

File: src/coinbase/address/wallet_address.ts (hunk 3 above)
Comment (marcin-cb) [reply src/coinbase/address/wallet_address.ts:365]: Yeah I'm planning on refactoring that a bit after our internal discussion. Will finalize things soon.

File: src/coinbase/staking_operation.ts (hunk 4 above)
Comment (marcin-cb) [reply src/coinbase/staking_operation.ts:72]: Oops, the case statements where suppose to be of type `StakingOperationStatusEnum` which is what comes from the model from the API call. Updating.

File: src/coinbase/types.ts [hunk 5]
```diff
@@ -689,6 +713,38 @@ export type CoinbaseConfigureFromJsonOptions = {
   basePath?: string;
 };
 
+/**
+ * CoinbaseExternalAddressStakeOptions type definition.
+ */
+export type CoinbaseExternalAddressStakeOptions = {
```
Comment (drohit-cb) [inline src/coinbase/types.ts:719]: unused

File: src/coinbase/types.ts [hunk 6]
```diff
@@ -689,6 +713,38 @@ export type CoinbaseConfigureFromJsonOptions = {
...
+  /**
+   * The amount to stake, unstake, or claim_stake for in a staking operation.
+   */
+  amount?: string;
+};
+
+/**
+ * CoinbaseWalletAddressStakeOptions type definition.
+ */
+export type CoinbaseWalletAddressStakeOptions = {
```
Comment (drohit-cb) [inline src/coinbase/types.ts:735]: can we not use this and instead take mode as an explicit param input ?

File: src/coinbase/wallet.ts [hunk 7]
```diff
@@ -289,6 +293,121 @@ export class Wallet {
...
+   * @throws {Error} if the default address is not found.
+   * @returns The staking operation after it's completed fully.
+   */
+  public async createStakingOperation(
+    amount: Amount,
+    assetId: string,
+    action: string,
+    timeoutSeconds = 60,
+    intervalSeconds = 0.2,
+    options: CoinbaseWalletAddressStakeOptions = { mode: StakeOptionsMode.DEFAULT },
```
Comment (drohit-cb) [inline src/coinbase/wallet.ts:396]: this func takes amount as input, and `CoinbaseWalletAddressStakeOptions` also has amount in it.
We should instead take mode as an explicit input and options separatey

File: src/coinbase/wallet.ts [hunk 8]
```diff
@@ -289,6 +293,121 @@ export class Wallet {
...
+   * @param assetId - The asset ID.
+   * @param startTime - The start time.
+   * @param endTime - The end time.
+   * @param format - The format to return the rewards in. (usd, native). Defaults to usd.
+   * @throws {Error} if the default address is not found.
+   * @returns The staking rewards.
+   */
+  public async stakingRewards(
+    assetId: string,
+    startTime = getWeekBackDate(new Date()),
```
Comment (drohit-cb) [inline src/coinbase/wallet.ts:368]: any reason why for external address we don't have this default but have it here ?
I like the idea of the defaults and we should maybe also add it to EA model

File: src/coinbase/wallet.ts (hunk 7 above)
Comment (drohit-cb) [reply src/coinbase/wallet.ts:396]: Also looks like we will take the route of renaming this to `createStakeOperation` ? and add the other 2 as well ?

File: src/coinbase/staking_operation.ts [hunk 9]
```diff
@@ -36,14 +37,77 @@ export class StakingOperation {
...
+    if (this.getWalletID() === undefined) {
+      throw new Error("cannot reload staking operation without a wallet ID.");
+    }
+    const result = await Coinbase.apiClients.stake!.getStakingOperation(
+      this.getWalletID()!,
+      this.getAddressID(),
+      this.getID(),
+    );
+    this.model = result?.data;
+    this.transactions = [];
```
Comment (drohit-cb) [inline src/coinbase/staking_operation.ts:105]: should we empty the transactions and then append to only when `result?.data.transactions` is non-empty to be safe ?

File: src/coinbase/types.ts [hunk 10]
```diff
@@ -708,6 +764,14 @@ export enum StakeOptionsMode {
   NATIVE = "native",
 }
 
+export enum StakingOperationStatus {
```
Comment (drohit-cb) [inline src/coinbase/types.ts:767]: do we need this or is `StakingOperationStatusEnum` is good enough ?

File: src/coinbase/wallet.ts (hunk 7 above)
Comment (marcin-cb) [reply src/coinbase/wallet.ts:394]: Yeah I added the refactor a few moments ago. They will be called `createStake` , `createUnstake` , and `createClaimStake`.

So based on your first comment, is this how you're suggesting the main function signature to look:
```
public async createStake(
    amount: Amount,
    assetId: string,
    mode = StakeOptionsMode.DEFAULT,
    timeoutSeconds = 60,
    intervalSeconds = 0.2,
    options: CoinbaseWalletAddressStakeOptions = {},
  ): Promise<StakingOperation> 
```
@drohit-cb

File: src/coinbase/wallet.ts (hunk 8 above)
Comment (marcin-cb) [reply src/coinbase/wallet.ts:368]: Yeah I'm in favor of adding these default to the EA model too.

File: src/coinbase/staking_operation.ts (hunk 9 above)
Comment (marcin-cb) [reply src/coinbase/staking_operation.ts:105]: Are you thinking about the scenario where we might get back an empty response and end up wiping the transactions mistakenly?

File: src/coinbase/types.ts (hunk 10 above)
Comment (marcin-cb) [reply src/coinbase/types.ts]: Yeah thought about it and does seem like there's any benefit to having an internal status representation. Removed it and put back the old `getStatus` function.

File: src/coinbase/wallet.ts (hunk 7 above)
Comment (drohit-cb) [reply src/coinbase/wallet.ts:394]: mode and options will look like this https://github.com/coinbase/coinbase-sdk-nodejs/blob/master/src/coinbase/address/external_address.ts#L32-L33

Comment (drohit-cb) [reply src/coinbase/wallet.ts:394]: and the order of params can be `amount, assetID, mode, options, timeoutSeconds, intervalSeconds`

File: src/coinbase/types.ts (hunk 6 above)
Comment (marcin-cb) [reply src/coinbase/types.ts]: Followed the suggestion on having mode param and removed this.

File: src/coinbase/wallet.ts (hunk 7 above)
Comment (marcin-cb) [reply src/coinbase/wallet.ts]: refactored based on the suggestion.

File: src/coinbase/staking_operation.ts (hunk 9 above)
Comment (drohit-cb) [reply src/coinbase/staking_operation.ts:92]: Yes

File: src/coinbase/address/wallet_address.ts [hunk 11]
```diff
@@ -340,4 +352,251 @@ export class WalletAddress extends Address {
...
+    );
+
+    // NOTE: Staking does not yet support server signers at this point.
+    await stakingOperation.sign(this.key!);
+    for (let i = 0; i < stakingOperation.getTransactions().length; i++) {
+      const transaction = stakingOperation.getTransactions()[0];
+      if (!transaction.isSigned()) {
+        continue;
+      }
+      stakingOperation = await this.broadcastStakingOperationRequest(
```
Comment (drohit-cb) [inline src/coinbase/address/wallet_address.ts:485]: do we need to track if a tx was previously broadcasted - if yes - then ignore ? Don't see any such handling within `broadcastStakingOperationRequest`

File: src/coinbase/address/wallet_address.ts [hunk 12]
```diff
@@ -340,4 +352,251 @@ export class WalletAddress extends Address {
...
+   * @returns The created staking operation.
+   */
+  private async createStakingOperationRequest(
+    amount: Amount,
+    assetId: string,
+    action: string,
+    mode: StakeOptionsMode = StakeOptionsMode.DEFAULT,
+    options: { [key: string]: string } = {},
+  ): Promise<StakingOperation> {
+    if (new Decimal(amount.toString()).lessThanOrEqualTo(0)) {
```
Comment (drohit-cb) [inline src/coinbase/address/wallet_address.ts:551]: this feels like an unnecessary hard check deep in the stack..
can the amount specific checks happen higher up like in the `validateCanClaimStake` etc funcc ?

File: src/coinbase/address/wallet_address.ts (hunk 11 above)
Comment (marcin-cb) [reply src/coinbase/address/wallet_address.ts:485]: There is a check to see if the transaction is signed above this call. The assumption is that if it's signed id' be broadcasted already. Would this break the dedicated ETH flow?

Comment (drohit-cb) [reply src/coinbase/address/wallet_address.ts:485]: Got it - Make sense. I think let's roll ahead with this and I can see if something needs to be addressed later for dedicated eth staking.

File: src/coinbase/address/wallet_address.ts [hunk 13]
```diff
@@ -340,4 +352,251 @@ export class WalletAddress extends Address {
...
+  /**
+   * Lists the staking rewards for the address.
+   *
+   * @param assetId - The asset ID.
+   * @param startTime - The start time.
+   * @param endTime - The end time.
+   * @param format - The format to return the rewards in. (usd, native). Defaults to usd.
+   * @returns The staking rewards.
+   */
+  public async stakingRewards(
```
Comment (drohit-cb) [inline src/coinbase/address/wallet_address.ts:516]: can this also be moved to the parent class ?

Comment (marcin-cb) [reply src/coinbase/address/wallet_address.ts:516]: Yeah can do.

File: src/coinbase/address/external_address.ts [hunk 14]
```diff
@@ -6,9 +6,7 @@ import { Asset } from "../asset";
 import { StakingOperation } from "../staking_operation";
```
Comment (drohit-cb) [inline src/coinbase/address/external_address.ts:6]: need to remove stakingRewards from here also ?

Comment (marcin-cb) [reply src/coinbase/address/external_address.ts:6]: yeah good call.

**Review by drohit-cb — APPROVED** [review]

**Pull request closed without merging**